andreeapetcu/assignment3/TrieNode.java

```diff
@@ -0,0 +1,208 @@
+  /* nodes with the same value but different positions are treated as different nodes;
+   * thus, the child with the given value consists of an array of nodes with identical values
+   * but different positions
+   */
+  private ArrayList<TrieNode>[] children;
+  boolean isWord;
+
+  // used to create the root only
+  public TrieNode(){
+    children = new ArrayList[LETTERS];
+  }
+
+  public TrieNode (char value) {
+    assert value >= 'a' && value <= 'z';
+    this.value = value;
+    children = new ArrayList[LETTERS];
+  }
+
+  public TrieNode (char value, int row, int column) {
+    assert value >= 'a' && value <= 'z' && row >= 0 && column >= 0;
+    this.value = value;
+    this.row = row;
+    this.column = column;
+    children = new ArrayList[LETTERS];
+  }
+
+  /**
+   * Adds a new word into the trie.
+   *
+   * @param word the word to be added
+   * @throws IllegalArgumentException if the word to be inserted is null
+   */
+  public void insertWord (String word) {
+    if (word == null) {
+      throw new IllegalArgumentException("null input");
+    }
+    
+    if (word.isEmpty()) {
+      return;
+    }
+
+    char letter = word.charAt(0);
+    ArrayList<TrieNode> child = children[getIndexOfCharacter(letter)];
+    if (child == null) {
+      children[getIndexOfCharacter(letter)] = new ArrayList<>();
+      children[getIndexOfCharacter(letter)].add(new TrieNode(letter));
+    }
+
+    if (word.length() > 1) {
+      children[getIndexOfCharacter(letter)].get(0).insertWord(word.substring(1));
+    } else {
+      children[getIndexOfCharacter(letter)].get(0).isWord = true;
+    }
+  }
+
+  /**
+   * Checks if the given string is a word in the trie.
+   *
+   * @param word the word whose existence in trie is checked
+   * @return true if the trie contains the word, false otherwise
+   * @throws IllegalArgumentException if the element to be searched for is null
+   */
+  public boolean containsWord (String word) {
+    if (word == null) {
+      throw new IllegalArgumentException("null input");
+    }
+
+    if (word.length() > 0) {
+      char letter = word.charAt(0);
+      ArrayList<TrieNode> child = children[getIndexOfCharacter(letter)];
+      if (child == null || !(child.get(0).value == letter)) {
+        return false;
+      } else {
+        return child.get(0).containsWord(word.substring(1));
+      }
+    }
+
+    return isWord;
+  }
+
+  /**
+   * Checks if the given string is a prefix in the trie.
+   *
+   * @param prefix the prefix whose existence in trie is checked
+   * @return true if the trie contains the prefix, false otherwise
+   * @throws IllegalArgumentException if the element to be searched for is null
+   */
+  public boolean containsPrefix (String prefix) {
+    if (prefix == null) {
+      throw new IllegalArgumentException("null input");
+    }
+
+    if (prefix.length() > 0) {
+      char letter = prefix.charAt(0);
+      ArrayList<TrieNode> child = children[getIndexOfCharacter(letter)];
```

Comment:
I don't understand the benefit of having a list and then always looking at element 0 in the list. Have you got confused here?